### PR TITLE
Raise a better error

### PR DESCRIPTION
**Pull request by nouiz**

### CODE REVIEW

**Comment by nouiz** [comment]
@hawkinsp

File: jax/_src/interpreters/partial_eval.py [hunk 1]
```diff
@@ -1857,6 +1857,10 @@ def process_primitive(self, primitive, tracers, params):
   def default_process_primitive(self, primitive, tracers, params):
     avals = [t.aval for t in tracers]
     out_avals, effects = primitive.abstract_eval(*avals, **params)
+    if not (isinstance(out_avals, (tuple,list)) == primitive.multiple_results):
```
Comment (yashk2810) [inline jax/_src/interpreters/partial_eval.py:1860]: can you split this apart into:

`if not isinstance(out_avals, (tuple, list)) and primitive.multiple_results:

File: jax/_src/interpreters/partial_eval.py [hunk 2]
```diff
@@ -1857,6 +1857,10 @@ def process_primitive(self, primitive, tracers, params):
...
+      raise ValueError(str(primitive) + ".abstract_eval() method should return"
```
Comment (yashk2810) [inline jax/_src/interpreters/partial_eval.py:1861]: Please add a test

Comment (nouiz) [reply jax/_src/interpreters/partial_eval.py:1861]: I don't find the current JAX custom call tests.
Where should I add a test?

Comment (nouiz) [reply jax/_src/interpreters/partial_eval.py:1861]: I added a test.

File: jax/_src/interpreters/partial_eval.py (hunk 1 above)
Comment (yashk2810) [reply jax/_src/interpreters/partial_eval.py:1860]: Can't you just replace `==` with `and`?

Comment (nouiz) [reply jax/_src/interpreters/partial_eval.py:1860]: No. Here is the table of those two operation. When both parameter are false, the results isn't the one we need.
```
              , ==, and
true, true,  true, true
true, false, false, false
false, true, false, false 
false, false, true, false
```

Comment (jakevdp) [reply jax/_src/interpreters/partial_eval.py:1860]: I would leave a comment here saying you're effectively using `==` as an XOR.

File: jax/_src/interpreters/partial_eval.py [hunk 3]
```diff
@@ -1857,6 +1857,10 @@ def process_primitive(self, primitive, tracers, params):
...
+                      " a tuple or a list if it return multiple results,"
+                      " otherwise it shouldn't return a tuple or a list.")
```
Comment (yashk2810) [inline jax/_src/interpreters/partial_eval.py:1863]: remove the last sentence or make it clear what it should return.

File: jax/_src/interpreters/partial_eval.py (hunk 1 above)
Comment (nouiz) [reply jax/_src/interpreters/partial_eval.py:1861]: Good idea. I said it is a not xor.

File: jax/_src/interpreters/partial_eval.py [hunk 4]
```diff
@@ -1857,6 +1857,11 @@ def process_primitive(self, primitive, tracers, params):
   def default_process_primitive(self, primitive, tracers, params):
     avals = [t.aval for t in tracers]
     out_avals, effects = primitive.abstract_eval(*avals, **params)
+    # == serve as a "not xor" here.
+    if not (isinstance(out_avals, (tuple,list)) == primitive.multiple_results):
+      raise ValueError(str(primitive) + ".abstract_eval() method should return"
+                       " a tuple or a list if " + str(primitive) +
```
Comment (yashk2810) [inline jax/_src/interpreters/partial_eval.py:1863]: Use f-strings instead of `+`?

**Review by yashk2810 — APPROVED** [review]

**Pull request merged**